File: src/api/java/BitVecNum.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    BitVecNum.java

Abstract:

Author:

    @author Christoph Wintersteiger (cwinter) 2012-03-15

Notes:
    
**/

package com.microsoft.z3;

import java.math.BigInteger;

/**
 * Bit-vector numerals
 **/
public class BitVecNum extends BitVecExpr
{
    /**
     * Retrieve the int value.
     * 
     * @throws Z3Exception
     **/
    public int getInt()
    {
        Native.IntPtr res = new Native.IntPtr();
        if (!Native.getNumeralInt(getContext().nCtx(), getNativeObject(), res)) {
            throw new Z3Exception("Numeral is not an int");
        }
        return res.value;
    }

    /**
     * Retrieve the 64-bit int value.
     * 
     * @throws Z3Exception
     **/
    public long getLong()
    {
        Native.LongPtr res = new Native.LongPtr();
        if (!Native.getNumeralInt64(getContext().nCtx(), getNativeObject(), res)) {
            throw new Z3Exception("Numeral is not a long");
        }
        return res.value;
    }

    /**
     * Retrieve the BigInteger value.
     **/
    public BigInteger getBigInteger()
    {
        return new BigInteger(this.toString());
    }

    /**
     * Returns a decimal string representation of the numeral.
     **/
    @Override
    public String toString()
    {
        return Native.getNumeralString(getContext().nCtx(), getNativeObject());
    }

    /**
     * Returns a binary string representation of the numeral.
     **/
    public String toBinaryString()
    {
        return Native.getNumeralBinaryString(getContext().nCtx(), getNativeObject());
    }

    BitVecNum(Context ctx, long obj)
    {
        super(ctx, obj);
    }
}

File: src/api/dotnet/BitVecNum.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    BitVecNum.cs

Abstract:

    Z3 Managed API: BitVec Numerals

Author:

    Christoph Wintersteiger (cwinter) 2012-03-20

Notes:
    
--*/
using System.Diagnostics;
using System;

#if !FRAMEWORK_LT_4
using System.Numerics;
#endif

namespace Microsoft.Z3
{
    /// <summary>
    /// Bit-vector numerals
    /// </summary>
    public class BitVecNum : BitVecExpr
    {
        /// <summary>
        /// Retrieve the 64-bit unsigned integer value.
        /// </summary>
        public UInt64 UInt64
        {
            get
            {
                UInt64 res = 0;
                if (Native.Z3_get_numeral_uint64(Context.nCtx, NativeObject, ref res) == 0)
                    throw new Z3Exception("Numeral is not a 64 bit unsigned");
                return res;
            }
        }

        /// <summary>
        /// Retrieve the int value.
        /// </summary>
        public int Int
        {
            get
            {
                int res = 0;
                if (Native.Z3_get_numeral_int(Context.nCtx, NativeObject, ref res) == 0)
                    throw new Z3Exception("Numeral is not an int");
                return res;
            }
        }

        /// <summary>
        /// Retrieve the 64-bit int value.
        /// </summary>
        public Int64 Int64
        {
            get
            {
                Int64 res = 0;
                if (Native.Z3_get_numeral_int64(Context.nCtx, NativeObject, ref res) == 0)
                    throw new Z3Exception("Numeral is not an int64");
                return res;
            }
        }

        /// <summary>
        /// Retrieve the int value.
        /// </summary>
        public uint UInt
        {
            get
            {
                uint res = 0;
                if (Native.Z3_get_numeral_uint(Context.nCtx, NativeObject, ref res) == 0)
                    throw new Z3Exception("Numeral is not a uint");
                return res;
            }
        }

#if !FRAMEWORK_LT_4
        /// <summary>
        /// Retrieve the BigInteger value.
        /// </summary>
        public BigInteger BigInteger
        {
            get
            {
                return BigInteger.Parse(this.ToString());
            }
        }
#endif

        /// <summary>
        /// Returns a decimal string representation of the numeral.
        /// </summary>        
        public override string ToString()
        {
            return Native.Z3_get_numeral_string(Context.nCtx, NativeObject);
        }

        /// <summary>
        /// Returns a binary string representation of the numeral.
        /// </summary>        
        public string ToBinaryString()
        {
            return Native.Z3_get_numeral_binary_string(Context.nCtx, NativeObject);
        }

        #region Internal
        internal BitVecNum(Context ctx, IntPtr obj) : base(ctx, obj) { Debug.Assert(ctx != null); }
        #endregion
    }
}
